TESTS/concurrent/Comms/Comms.cpp

```diff
@@ -0,0 +1,217 @@
+
+    // write to EEPROM
+    num_written = memory.write(address,Test_String,TEST_STRING_MAX);
+
+    // read back from EEPROM
+    num_read = memory.read(address,(char *)read_string,TEST_STRING_MAX);
+
+    // test for equality
+    TEST_ASSERT_EQUAL_STRING_MESSAGE(Test_String,(char *)read_string,"String read does not match the string written"); // test that strings match
+    TEST_ASSERT_EQUAL_MESSAGE(num_written,num_read,"Number of bytes written does not match the number of bytes read");
+    DEBUG_PRINTF("\r\n****\r\n I2C TEST:\r\n Address = `%d`\r\n Num Bytes Written = `%d` \r\n Num Bytes Read = `%d` \r\n String written = `%s` \r\n String read = `%s` \r\n****\r\n",address,num_written,num_read,Test_String,read_string);
+    osSignalSet(Multi_Thread_ID, 0x1);                                // signal completion of thread
+}
+
+
+// test SPI API by writing and reading from SD card
+void test_SPI()
+{
+    // initialze SD hardware, filesystem, and variables
+    SDBlockDevice sd(MBED_CONF_APP_SPI_MOSI, MBED_CONF_APP_SPI_MISO, MBED_CONF_APP_SPI_CLK, MBED_CONF_APP_SPI_CS);
+    FATFileSystem fs("sd", &sd);
+    sd.init();
+    fs.mount(&sd);
+    FILE *File = fopen("/sd/test_sd_w.txt", "w");
+    TEST_ASSERT_MESSAGE(File != NULL,"SD Card is not present. Please insert an SD Card.");
+    volatile char read_string[TEST_STRING_MAX] = {0};
+
+    // write data to SD card
+    int error = fprintf(File, Test_String);                           // write data
+    TEST_ASSERT_MESSAGE(error > 0,"Writing file to sd card failed");  // error checking       
+    fclose(File);                                                     // close file on SD
+
+    // read data from SD card
+    File = fopen("/sd/test_sd_w.txt", "r");
+    fgets((char *)read_string,TEST_STRING_MAX,File);                  // read string from the file
+
+    // test for equality
+    TEST_ASSERT_EQUAL_STRING_MESSAGE(Test_String,(char *)read_string,"String read does not match the string written"); // test that strings match
+    DEBUG_PRINTF("\r\n****\r\n SPI TEST:\r\n String written = `%s` \r\n String read = `%s` \r\n****\r\n",Test_String,read_string);
+
+    // close, unmount, and deinitialize
+    fclose(File);                                                     // close file on SD
+    fs.unmount();
+    sd.deinit();
+    osSignalSet(Multi_Thread_ID, 0x2);                                // signal completion of thread
+}
+
+
+// test I2C and SPI APIs concurrently in multiple threads
+void test_multiple_threads()
+{
+    Multi_Thread_ID = Thread::gettid();                               // update thread id for this thread
+    Thread_I2C.start(callback(test_I2C));                             // kick off threads
+    Thread_SPI.start(callback(test_SPI));                             // kick off threads
+    wait(0.1);                                                        // allow time for debug print statements to complete.

```

Comment:
Yes, both threads could complete their work, and the test could exit before the debug print finishes its printing.